### ISSUE TITLE
Supporting ChangeMessageVisibility heartbeat

### ISSUE THREAD

**Issue by kcasamento** [bbc/sqs-consumer#171]
Question/Enhancement:  Or is it possible to get a reference to the underlying queue object inside the handleMessage callback and/or is there some event I can attached to that will be called on an interval that allows me to manually call "ChangeMessageVisibility"?  Or is this functionality natively supported through config options?

Reference:
https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_ChangeMessageVisibility.html

**Comment by niklasR** [comment]
When do you want to call it?
As per the [README](https://github.com/bbc/sqs-consumer/blob/master/README.md), you can set your own config values for `visibilityTimeout` and `terminateVisibilityTimeout`.

**Comment by kcasamento** [comment]
On an interval after a message has been dequeued. The idea here is I have a queue setup to process file uploads and the time it takes to process is dependent on file size. So to set a default visibility timeout of 2 minutes and then say every 1.5 minutes if the file is still uploading being able to update the visibility timeout for another 2 minutes etc.

**Comment by kcasamento** [comment]
If it's of any interest I forked the code and wrote what I need, but was looking for something to this effect per the guidance from AWS doc [here](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/sqs-examples-managing-visibility-timeout.html)

```javascript
// Create Consumer with 2 new parameters
const app = Consumer.create({
            queueUrl,
            heartbeatIntervalSeconds: 20,
            heartbeatVisibilityTimeoutSeconds: 20,
            handleMessage: async (message) => {
     
                // Unpredictably long running task
                await this._handleMessage(message);
                
            }});
```

```javascript


function createHeartbeat(intervalSeconds: number, fn: Function, ...args: any): NodeJS.Timeout {

  let heartbeat: NodeJS.Timeout;

  heartbeat = 
    setInterval( () => fn.apply(this, args), intervalSeconds * 1000);  
  heartbeat.unref(); 

  return heartbeat;

}
private async executeHandler(message: SQSMessage): Promise<void> {
    let timeout;
    let pending;
    let heartbeat;

    try {

        if (this.heartbeatInterval > 0) {

          heartbeat = createHeartbeat(
            this.heartbeatInterval, 
            this.handleHeartbeat, 
            this.queueUrl, 
            this.heartbeatVisibilityTimeout, 
            message);

        }

        if (this.handleMessageTimeout) {
            [timeout, pending] = createTimeout(this.handleMessageTimeout);
            await Promise.race([
            this.handleMessage(message),
            pending
            ]);
        } else {
            await this.handleMessage(message);
        }
    } catch (err) {
      if (err instanceof TimeoutError) {
        err.message = `Message handler timed out after ${this.handleMessageTimeout}ms: Operation timed out.`;
      } else {
        err.message = `Unexpected message handler failure: ${err.message}`;
      }
      throw err;
    } finally {
      clearTimeout(timeout);
      clearInterval(heartbeat);
    }
  }

private async handleHeartbeat(queueUrl: string, increaseTimeout: number, message: SQSMessage): Promise<any> {

    return new Promise( (resolve, reject) => {

        const visibilityParams = {
            QueueUrl: queueUrl,
            ReceiptHandle: message.ReceiptHandle,
            VisibilityTimeout: increaseTimeout
        };
    
        this.sqs.changeMessageVisibility(visibilityParams, (err, data) => {
            if (err) {
              console.log("Delete Error", err);
              resolve(err);
            } else {
              console.log("Timeout Changed", data);
              resolve(data);
            }
        });

    });
    
  }

```

**Comment by cdhowie** [comment]
Are there any plans to support this feature? This is important for long-running jobs.

Consider the case where the default visibility timeout is one minute, but jobs might take over an hour. It's important that the message is continuously re-hidden while the job continues to run.

It would be a bad idea to use a default visibility timeout of an hour or more, because then if a worker dies unexpectedly (hardware/network failure) we want the job to become available to other workers quickly.

**Comment by machulav** [comment]
At the moment we have the same problem with long-running jobs and this feature could be very helpful in this case!

**Comment by naren8642** [comment]
@kcasamento your solutions like a good way to address this. Why don't you make a pull request?

**Comment by mjgp2** [comment]
I would be up for potentially helping with a fork @kcasamento !

**Comment by pwrnrd** [comment]
@mjgp2 and @kcasamento I created an update of this package (SQS Consumer + SQS Producer) in Typescript. If you want Typescript support I can set-up a project and publish my work. I'm not interested in maintaining such a project alone, but if we can work together on that then that would be fine. In that case, let me know, I can set-up the repo's this weekend.

**Comment by mjgp2** [comment]
We are Typescript users, so would be interested for sure!

**Comment by pwrnrd** [comment]
Ok I've set-up a repo: https://github.com/pwrnrd/aws-sqs-consumer . Accepting PRs.

**Comment by mjgp2** [comment]
Awesome, will share with the team

**Comment by nspragg** [comment]
See `heartbeatInterval` (released in 5.5.0)
